### ISSUE TITLE
Auto create pull requests after releases for automatic updates

### ISSUE THREAD

**Issue by mathias-baumann-sociomantic** [sociomantic-tsunami/neptune#70]
Write a program/script that auto-creates pull request to all repos using the library that was just released.
It would need to check if the repo is using the library at all and also which version of it.

When there are several maintained version, create an update PR to the matching major version.
If the repo uses an unmaintained version, update it to the oldest maintained.

Optionally, create a github-project that includes all those PRs.

This would be some extra program called by `neptune-release` on request.

Open question: What if there are still existing PRs from earlier releases?

**Comment by mathias-baumann-sociomantic** [comment]
Relevant: https://mdswanson.com/blog/2011/07/23/digging-around-the-github-api-take-2.html

**Comment by mathias-lang-sociomantic** [comment]
Pull requests updating to what ? Patch and minor alike, or just patch ? Also how do you deal with libraries ?

**Comment by mathias-baumann-sociomantic** [comment]
When new version of a library is released, this program would search all projects that use this library and create a PR to each of them, updating to the just released version *IF* it is the same major version.

**Comment by mihails-strasuns-sociomantic** [comment]
I think automatic approach should only be enabled for patch releases. And minor ones only done by some explicit flag enabled manually. When doing latter libraries should be ignored too.

> Open question: What if there are still existing PRs from earlier releases?

Make it create PR with fixed title prefix like "[neptune auto-upgrade PR] ..." and check if such PR exists before creating new one. If it exists, update/replace.

**Comment by mathias-baumann-sociomantic** [comment]
> I think automatic approach should only be enabled for patch releases. 

why?

**Comment by mihails-strasuns-sociomantic** [comment]
Because those are only ones that can be merged with more or less zero thought. Even for minor updates app maintainers would want to do at least limited production testing - I feel that bothering them each single release about it is too much annoyance.

Maybe only doing automated PR if minor is 2+ versions old (== unsupported)?

**Comment by leandro-lucarella-sociomantic** [comment]
We'll probably have to try and see, so I'd suggest to make all that configurable, so it could be tuned based on how much we bother downstream projects. Also, this might be something that downstream project could configure via `.neptune.yml`, so a project might want to receive PRs for every release while other might just want to patch releases.

**Comment by leandro-lucarella-sociomantic** [comment]
[slightly off-topic] 

About this, I find myself very often trying to do mass update of projects for other stuff (like changing `.editorconfig` or upgrading the cachalot image versions). The overlap between this and what this issue is about seems to be very high, so I wonder if we could make the tool general enough to support this too.

I'm thinking of providing two options, one would be to provide a diff/patch file to apply and another would be just a script. The patch then will be applied to all projects (or the script run for all projects), so the tool will just fetch all the relevant projects, apply the script/patch, commit and make a PR.

If this can be done purely via GitHub API (as suggested by the link you posted), that would be awesome!

**Comment by leandro-lucarella-sociomantic** [comment]
Following the idea of keeping track of PRs, we could add an identifier, for example: `neptune-mass-pr --script convert-spaces.sh --id spaces --subject "Fix spaces" ` will create a PR named: `[neptune-mass-pr spaces] Fix spaces`. If I find an issue, I fix the `convert-spaces.sh` script and re run: `neptune-mass-pr --script convert-spaces.sh --id spaces --subject "Fix spaces, second try" `, which will look for a PR with subject starting with `[neptune-mass-pr spaces] ` and update it (or create a new one if it doesn't exist) and change the subject to: `[neptune-mass-pr spaces] Fix spaces, second try`. For a library update `--id`  can be `ocean-v6.2.x` for example (if we would want to reuse the same PR for patches but not for other releases (or just `ocean` if we only want to keep one mass-pr alive at the time).

**Comment by mathias-baumann-sociomantic** [comment]
We can also fetch the branch name used for a PR. Maybe easier?

**Comment by mathias-baumann-sociomantic** [comment]
Regarding whether to do it only for patches or also for minors: We shouldn't forget the advantage that creating those PRs is also a kind of test for all us to see if we broke anything. If a PR fails we know that we apparently did a SemVer incompatible change.

**Comment by mathias-baumann-sociomantic** [comment]
A question that is coming up for me currently: Should the branches for the PRs created in a fork or in the main-repo? Technically it would be much easier to do it in the main repo.

Otherwise I probably would add a `--setup` flag which would fork all the repos and of course if a new repo is created this would need to be done again.. as it is not too easy to check if we already have a fork I wouldn't want to do that every single time (it's possible, but I try to get ALL info I need in one call and adding it there means I get the list of ALL forks for ALL repos, so lots of data.)

**Comment by mathias-baumann-sociomantic** [comment]
Okay, that wasn't exactly correct. I CAN get the list of all repos for one specific user, so at least that is realistic. Still, it does require more setup and care if we go the fork way

**Comment by gavin-norman-sociomantic** [comment]
A small point on this that just came to my mind: any release tagged `+breaking` shouldn't have an auto-PR, because it will by definition fail. Right?

**Comment by mathias-baumann-sociomantic** [comment]
No, just because one part of the API changes doesn't mean that part was used in the actual project.

Another thing to consider making your small point a bit more a major point: If you are on version `v1.2.3` and want to update to version `v1.5.6` but there was a `v1.3.3+breaking`, every version AFTER the breaking one must also be considered breaking.

**Comment by mathias-baumann-sociomantic** [comment]
So basically, there would be no auto-prs anymore until someone manually updated to or above the breaking version. Is that really desirable?

I'd rather add some warning notes in the PR that someone should manually look at the breaking change.

**Comment by gavin-norman-sociomantic** [comment]
>No, just because one part of the API changes doesn't mean that part was used in the actual project.

True. I suppose we could just do it and see whether the project fails.

**Comment by leandro-lucarella-sociomantic** [comment]
> A question that is coming up for me currently: Should the branches for the PRs created in a fork or in the main-repo? Technically it would be much easier to do it in the main repo.

I think that since this is an automated thing, and is just "infra" stuff, and there shouldn't be many PRs at a time, it might be OK to do in the main repo. I think it even makes more sense to do in the main repo, I'm just a bit concerned about branch pollution in the main repo, but again, if PRs are reused when new versions are out, there should probably be only one PR pending at the time.

**Comment by mathias-lang-sociomantic** [comment]
How would that work with permissions ? Wouldn't the person running the tool have to be an admin on all repos then ?

**Comment by mathias-baumann-sociomantic** [comment]
@leandro-lucarella-sociomantic you forget that any given project might use several submodules. Well, very likely even. Makd, ocean, swarm, dive.. each could have a PR

**Comment by mihails-strasuns-sociomantic** [comment]
I am for forking with dedicated bot account.

**Comment by leandro-lucarella-sociomantic** [comment]
> @leandro-lucarella-sociomantic you forget that any given project might use several submodules. Well, very likely even. Makd, ocean, swarm, dive.. each could have a PR

Yeah, one per submodule, it's still a bounded amount of time. I don't have a strong opinion, it just seems simpler to use the blessed repo.
